File: backend/python/PolledControllerWrapper.py
#!/usr/bin/python

from IRobot import IRobot
from IRobotController import IRobotController
from Point import Point

class PolledControllerWrapper(IRobotController,object):
    def __init__(self,ctlr):
        self.controller = ctlr
        self.robot = IRobot()
        #print(self.controller,"\n",hasattr(self.controller,'controlRobot'))
        #print(self.controller.controlRobot)
        #self.controlRobot = None
        #self.ctlrReset = None
        #if hasattr(ctlr,'controlRobot'):
        #    self.controlRobot = getattr(ctlr,'controlRobot')
        #if hasattr(ctlr, 'reset'):
        #    self.ctlrReset = getattr(ctlr,'reset')

    def start(self):
        while (self.robot.getLocation() != self.robot.getTargetLocation()):
            #self.controlRobot(self.robot)
            self.controller.controlRobot(self.controller,self.robot)
            self.robot.jsondump()

    def setRobot(self, robot):
        self.robot = robot
    
    def reset(self):
        self.controller.reset(self.controller)
        self.robot.reset()

    def getDescription(self):
        return "Polled Controller"

File: backend/java/src/PolledControllerWrapper.java
import java.lang.reflect.Method;

public class PolledControllerWrapper implements IRobotController {
    IRobot robot;
    Object pollObj;
    Method controlRobot;
    Method reset;

    public PolledControllerWrapper(Object obj) {
        pollObj = obj;
        // Attempt to find a controlRobot method in the given Class
        try {
            controlRobot = obj.getClass().getMethod("controlRobot", new Class[] {IRobot.class});
        } catch (Exception e) {
            e.printStackTrace();
        }

        // Attempt to find a reset method in the given Class
        try {
           reset = obj.getClass().getMethod("reset", (Class[]) null);
        } catch (Exception e) {
        }
    }

    public void start() throws ResetException {
        // Invoke the controlRobot method until the finish location is reached
        while(!robot.getLocation().equals(robot.getTargetLocation())) {
            try {
                controlRobot.invoke(pollObj, new Object[] {robot});
                robot.jsondump();
            } catch (ResetException re) {
                throw re; // If the controller is reset, propogate the ResetException upwards to be handled elsewhere
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    public void reset() {
        // If the controller has a reset function, call it before resetting normally
        if(reset != null){
            try {
                reset.invoke(pollObj, (Object[]) null);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
        robot.reset();
    }

    public void setRobot(IRobot robot) {
        this.robot = robot;
    }

    public String getDescription() {
       String tag = "Polled Controller";
       if (reset != null)
          tag += " with reset()";
       else
          tag += " w/o reset()";

       return tag;
    }
}